[DOCS] Draft NLP model deployment (#1905)
elastic/stack-docs: docs/en/stack/ml/nlp/ml-nlp-deploy-models.asciidoc

[[ml-nlp-deploy-models]]

> = Import and deploy trained models

Author: Would you prefer "Select and deploy..." in that case, or just "Deploy..."?

> = Import and deploy trained models

Reviewer: Good question. I guess I would just say "Deploy..." for now. Also since a customer may train their own model. So then we would have to say, "Train or select and then deploy" 😆

:keywords: {ml-init}, {stack}, {nlp}
:description:  To make trained models available for inference, you must import \
and deploy them in {es}.

If you want to perform {nlp} tasks in your cluster, you must import and deploy
an appropriate trained model. There are {es} APIs and tooling support in eland
to prepare and manage models. 

. <<ml-nlp-select-model,Choose a trained model.>>
. <<ml-nlp-upload-model,Upload the trained model and vocabulary.>>
. <<ml-nlp-deploy-model,Deploy the model in your cluster.>>

[discrete]
[[ml-nlp-select-model]]
== Select a trained model

Per <<ml-nlp-overview>>, there are multiple ways that you can use NLP features
within the {stack}. After you determine which type of NLP task you want to
perform, you must choose an appropriate trained model. You can either create the
model yourself or find one that meets your needs.

IMPORTANT: The {stack-ml-features} support transformer models that conform to
the standard BERT model interface and use the WordPiece tokenization algorithm.

The simplest method is to use a model that has already been fine-tuned for the
type of analysis that you want to perform. For example, there are models and 

> data sets available for specific NLP tasks in

Author: Done!

https://huggingface.co/models[Hugging Face].

[discrete]
[[ml-nlp-upload-model]]
== Upload the trained model and vocabulary

> After you find or create a model, you must upload it and its tokenizer

Author: Good point.  I've replaced all "uploads" with "imports".

vocabulary to your cluster. 

Trained models must be in a TorchScript representation for use with

> {stack-ml-features}. To use scripts that format and upload the necessary

Reviewer: Rephrase to something like:
To use scripts that convert existing Hugging Face transformer models to their TorchScript representations, and import them with their tokenizer vocabularies, refer to https://github.com/elastic/eland#nlp-with-pytorch. If you have trained a model yourself using PyTorch directly, you will be responsible for first creating the TorchScript representation of the model. Once you have a TorchScript representation, you can use eland to import the necessary artifacts into Elasticsearch: To use scripts that convert existing Hugging Face transformer models to their TorchScript representations, and import them with their tokenizer vocabularies, refer to https://github.com/elastic/eland#nlp-with-pytorch.

Author: Thanks!  I've reworked it in 04a2fce3f59f90c5ee3fdeeef1a2705b187fe070

artifacts, refer to https://github.com/elastic/eland#nlp-with-pytorch

Alternatively, you can use the
{ref}/put-trained-model-vocabulary.html[create trained model vocabulary API] and

> {ref}/put-trained-models.html[create trained models API] or

Reviewer: Why is this an "or"? I think it's an "and" 😃 You need to create the trained model vocabulary, put the trained model, and put all the model definition part.

Author: Sorry, I misunderstood that you created a model or created part of a model.  I've fixed that in 04a2fce3f59f90c5ee3fdeeef1a2705b187fe070

{ref}/put-trained-model-definition-part.html[create trained model definition part API].
// When you upload the model, it must be chunked and uploaded one chunk at a time. 

> //TBD Why? How?

Reviewer: This is true, that's why I'm not sure we should even mention the API's directly but just the eland scripts since they encapsulate the chunking, for example. We have to chunk the model since it is stored in parts in an ES index, and we can't store very large (1GB+) models as a single document.

Author: Thanks, for the details!  For now, I've qualified that paragraph with the recommendation to use eland.

//Since eland encapsulates this process in a single Python method, it is the recommended method.

[discrete]
[[ml-nlp-deploy-model]]
== Deploy the model in your cluster

After you upload the model and vocabulary, you can use {kib} to view and manage

> their deployment across your cluster. Alternatively, you can use the

Reviewer 2: Does it make sense to further specify the location?
```suggestion
their deployment across your cluster under **{ml-app}** > _Model Management_. Alternatively, you can use the
```

{ref}/start-trained-model-deployment.html[start trained model deployment API].

> When you deploy the model, it is loaded on all available {ml} nodes. You can

Reviewer: ```suggestion
When you deploy the model, it is allocated to all available {ml} nodes. You can
```

I don't know. I'm just trying to unify all of the verbiage still.

optionally specify the number of CPU cores it has access to on each node. In

> general, the total size of threading settings across all models on a node should

Reviewer: We need to reference the API that allows you to set the thread settings. There are also two kinds of threads to set. In the original doc, I think I reference the two types `model_thread` and `inference_threads`. It's really important that we explain the difference between these two settings, either here or in the API.

Author: Yes, I played with adding that content but it felt too low-level for the current state of this page.  I'll open a PR to add it to the API for now and we can revisit as we get more examples into these docs.

Author: I changed my mind and added some of those recommendations in 23b14318e4492cb09b4a848deae1a7d7e2d55937

Author: I have, however, made some other minor edits in https://github.com/elastic/elasticsearch/pull/81549

not exceed the number of physical CPU cores available on the node, minus one
(for non-inference operations). In {ecloud} environments, the core count is
virtualized CPUs (vCPUs) and this total size should typically be no more than
half the available vCPUs, minus one.

The model is loaded into memory in a native process that encapsulates

> libtorch, which is an underlying machine learning library for PyTorch.

Reviewer: Maybe?

```suggestion
`libtorch`, which is the underlying machine learning library of PyTorch.
```
